### DISCUSSION TITLE
Is there any present way to hack in Golang objects or functions to be available in the JSVM?

### DISCUSSION THREAD

**Discussion by dschissler** [pocketbase/pocketbase#3853] [Q&A]
I'm trying to figure out how to expose functions and structs from Golang to JSVM.  Would it work to pass in an object in the route context?  Are there any other hacks for bringing functions and middleware in from my custom code?

Part of the issue is that compiling a Golang code change on a RPI 4B takes over 10 seconds.  So I'm thinking that JSVM could be better.

**Comment by ganigeorgiev** [comment]
An alternative workaround to the route context could be to define your functions as struct methods and attach a value of it to `app.Cache().Set("utils", &yourUtilStruct)` (_in the next release `app.Cache()` is soft-deprecated in favor of `app.Store()`_).

```go
type Utils struct { /*  add fields if necessary but be aware of concurrent access */ }

func (u *Utils) SomeMethod() string {
    return "hello world"
}

app.Cache().Set("utils", &Utils{})
```

Then from JSVM-side you can access it by doing something like `$app.cache().get("utils").someMethod()`.

**Reply by dschissler** [reply]
That seems good.  I'm not marking it as an answer though since it's hopefully not permanent.

Perhaps later you could allow us to place these structs into a special singleton that will be automatically available in the JSVM prefixed by `$`?

The trick would be to not fire up the JSVM binding too soon.  It's probably best that this is done between `OnAfterServe` (although maybe it already is).  The reason is that I don't start getting my Web stuff loaded up until that event.

**Reply by dschissler** [reply]
You could even refactor the current bootstrap to use this setup method, and then just hand it off to us.

**Reply by dschissler** [reply]
@ganigeorgiev If you were to allow us to bind our own JSVN top-level variables then perhaps you could make them available as double dollar signs `$$` in order to prevent future collisions as you expand the built-ins.  This will make it clear what is custom and what is core. 

I'm thinking about down the road as an third-party ecosystem develops.  It could be very confusing for a new person figuring out what has been provided by a custom setup or example, and what is built-in.

Also overloading these built-ins with custom extensions seems like a total mind-fuck that shouldn't be allowed.  That means that using a new custom third-party extension could overload something that another extension relies upon.  Probably each key should only allowed to be set once in order to prevent incompatibility (by encouraging forced resolution of the collision).  It' would also be difficult to diagnose these sorts of problems due to it having to do with two different environments that someone would need to simultaneously understand.

Imagine someone overloading a url encoding function.  It should never happen.

**Comment by dschissler** [comment]
@ganigeorgiev If it seems confusing as to what my goal is then this will make it more clear.  I want to create a lot of utilities for exposure to the JSVM such that I can choose to just work in there if I want to.

I'm retooling my casbin echo middleware to be more tightly integrated with pocketbase, and also so that it's JSVM friendly with a auth middleware chainable builder.  I've integrated in optional fake-group memberships for unknown requests, unverified (email) user requests, and admin via the admin token.

So you can define what admin token users should be able to do.

```csv
# root can do everything
p, /*, @pb-admin (GET|POST|PUT|PATCH|DELETE|OPTIONS)
``` 

Here I define group inheritance, but you can do it however you want.

```csv
g, @pb-unverified, @pb-unknown
g, @normal, @pb-unverified
g, @power, @normal
g, @admin, @power
g, @super, @admin
```

resftul Keymatch2 seems to perfectly coincide with echo routes.

```csv
# invitation controller
p, /invitation/:id, @unknown, GET
p, /invitation, @normal, GET
p, /invitation/approve/:id1/:id2, @normal, GET
```

I'm going to make this very easy to setup, and with the builder it will bring in the adapters for embed fs, string, and normal file path.  For a serious app you'll maybe want to put the policy in embed (as I'm doing now), but for other apps that file can live in the data or hooks directory, which allows for a built-in live reload.

I already have a required definable function whose purpose is to return a group name that is used against the casbin policy.  It's basically impossible to create just 2-3 patterns that will work for everyone.  This will also work in JSVM because you've added nice query support there.

I'm thinking about a project name of `caspoc`, or `caspocket`.

I'm thinking about adding HTMX support as a way to learn it, and because it could be very nice for programming without a Javscript build tool.

I've created most of a linux camera route already, with just shutdown necessary (once you can fix Restart not firing onTerminate).  I want to bind more utilities for controlling Raspberry Pi relays, lights, etc.

So one thing that I noticed while adding structs to the JSVM through app.cache() is that there is a very narrow window for binding these things to be available in the top level of the JSVM, but also after the initial bootstrapping.  For that particular use case it felt like that another another event inbetween boostrap starting and ending would be useful.  There are ways to make it work, but it's awkward.

About the casbin code, I'm hoping that you can see how easy that it is to use, and that it would inspire you to bake some support for it into Pocketbase.
